src/cli/telemetry/telemetry.ts

```diff
@@ -0,0 +1,309 @@
+import { tryFn, hasDebug, readJson, hasVerbose } from './helpers';
+import { shouldTrack } from './shouldTrack';
+import { CompilerBuildResults, Config, PackageJsonData, TaskCommand } from 'src/declarations';
+import { readConfig, updateConfig } from '../ionic-config';
+import { isObject } from '@utils';
+import { getCompilerSystem, getCoreCompiler, getStencilCLIConfig } from '../state/stencil-cli-config';
+
+/**
+ * Used as the object sent to the server. Value is the data tracked.
+ */
+export interface Metric {
+  name: string;
+  timestamp: string;
+  source: 'stencil_cli';
+  value: TrackableData;
+}
+
+/**
+ * Used for deeper understanding of how people are using certain output targets
+ */
+interface OutputTargetOptions {
+  serviceWorker?: boolean;
+}
+
+/**
+ * The task to run in order to collect the duration data point.
+ */
+type TelemetryCallback = (...args: any[]) => void | Promise<void>;
+
+/**
+ * The model for the data that's tracked.
+ */
+export interface TrackableData {
+  yarn: boolean;
+  session_id: string;
+  component_count?: number;
+  arguments: string[];
+  targets: string[][] | OutputTargetOptions[][];
+  task: TaskCommand;
+  duration_ms: number;
+  packages: string[];
+  os_name: string;
+  os_version: string;
+  cpu_model: string;
+  typescript: string;
+  rollup: string;
+  system: string;
+  build: string;
+  stencil: string;
+}
+
+/**
+ * Used to within taskBuild to provide the component_count property.
+ * @param result The results of a compiler build.
+ */
+export async function telemetryBuildFinishedAction(result: CompilerBuildResults) {
+  const { sys, flags, logger, validatedConfig } = getStencilCLIConfig();
+  const tracking = await shouldTrack(flags.ci);
+
+  if (!tracking) {
+    return;
+  }
+
+  const details = sys.details;
+  const session_id = await getTelemetryToken();
+  const packages = await getInstalledPackages();
+  const targets = await getActiveTargets(validatedConfig.config);
+  const versions = getCoreCompiler()?.versions || { typescript: 'unknown', rollup: 'unknown' };
+  const component_count = Object.keys(result.componentGraph).length;
+
+  const data: TrackableData = {
+    session_id,
```

Comment:
Sorry, I missed this earlier. The `session_id` should be a part of `Metric` and be set on the message in `sendMetric()`.
```suggestion
```

Reply:
Ohhh! Got you.